### ISSUE TITLE
Detours via Nuget package ?

### ISSUE THREAD

**Issue by DrusTheAxe** [microsoft/Detours#124]
I need to use Detours in a project so the simplest answer is "use Detours' nuget package" - except that doesn't exist :-(

I tried adding Detours source into my project but for complicated reasons its fails and requires invasive changes to my project settings and/or structure

I could clone Detours source + makefiles and make a project whose sole purpose is to run make to build Detours' static lib for my project to consume but that's irksome (to live with as well as to setup - local builds, build pipelines, etc)

Is there a reason Detours offers no Nuget package?

**Comment by sylveon** [comment]
You can use the vcpkg package. NuGet is generally a poor package manager for native dependencies.

**Comment by bgianfo** [comment]
Yea, the vcpkg that @sylveon contributed is a good option.
See: https://github.com/microsoft/vcpkg/tree/master/ports/detours

**Comment by DrusTheAxe** [comment]
@sylveon Be that as it may Nuget has a certain currency in my process. An official nuget from the Detours team would be ideal

@bgianfo  I'm not familiar with vcpkg. How do I use that? Do I have to build something or can I point VS2019 at something to import it?

**Comment by sylveon** [comment]
If you follow the instructions in the vcpkg readme for setting it up and then run `vcpkg install detours:x64-windows` (replace x64 by the preferred architecture), it should be automatically picked up by Visual Studio under `<detours/detours.h>`

**Comment by DrusTheAxe** [comment]
Ping

Still using a forked copy of Detours source and still interested in prebuilt Nuget packages

**Comment by acepace** [comment]
I think I am using a nuget package and it works well enough. 4.0.1

**Comment by DrusTheAxe** [comment]
Ping

Any news?

**Comment by DrusTheAxe** [comment]
Ping. Any update?

**Comment by bgianfo** [comment]
No updates as of yet, sorry. 

@DrusTheAxe Do you or the WindowsAppSDK team have cycles to help contribute the Nuget creation automation? That would certainly help move the needle here.

**Comment by DrusTheAxe** [comment]
> No updates as of yet, sorry.
> 
> @DrusTheAxe Do you or the WindowsAppSDK team have cycles to help contribute the Nuget creation automation? That would certainly help move the needle here.

I don't anytime soon. Will ask around but folks are generally pretty booked with higher pri work

**Comment by stevenbrix** [comment]
FYI, there is this nuget package: https://www.nuget.org/packages/Detours

which to the nuget package manager looks like it's created by Microsoft:
![image](https://github.com/microsoft/Detours/assets/9649518/fb29db6f-91f8-449c-9fed-2824ee47767f)

but turns out it isn't:
![image](https://github.com/microsoft/Detours/assets/9649518/bb86be54-38d1-470d-9e51-73bf9c48ff3b)

kinda scary that it's this easy to spoof being microsoft

**Comment by amithegde** [comment]
I tried using the `vcpkg`, looks like the commands have changed, now we have to use the `add` command like this `vcpkg add port detours` howerver after doing necessary setup on VS 2022, I get this error: `detours.h(47,1): error C1189: #error:  Unknown architecture (x86, amd64, ia64, arm, arm64)`

Can instructions to set it up on VS be published on the readme file so that we don't have to spend an hour to figure out how to use this library? Also, publishing to nuget is way better

**Comment by sylveon** [comment]
NuGet kinda sucks for C++. AFAIK, it's still `vcpkg install detours`, from when I tested some weeks ago.

Your error most likely is cause you need to include windows.h first.

**Comment by amithegde** [comment]
> NuGet kinda sucks for C++. AFAIK, it's still `vcpkg install detours`, from when I tested some weeks ago.
> 
> Your error most likely is cause you need to include windows.h first.

I got the vcpkg working and setup detours, but when I sync the repo on another machine, I again need to setup vcpkg and some of the initialization files need to be regenerated. I felt this doesn't scale. May be I don't know how to make it work on multiple machines just with cloning the repo.

**Comment by sylveon** [comment]
You can include a vcpkg.json file in your project for it to automatically restore detours on build. Newer versions of VS come with vcpkg built-in.

**Comment by amithegde** [comment]
> You can include a vcpkg.json

I already have vcpkg.json file and following is the content of the file:
`{
  "dependencies": [
    "detours"
  ]
}
`
I synced the repo on another machine, tried to build it on VS using `Ctrl+Shift+B` and I see this error. `error C1083: Cannot open include file: 'detours/detours.h': No such file or directory`

I don't see how the vckpg setup can be made portable. I don't want to manuvally configure vcpkg on every machine, I jsut want to clone the repo and continue developing. That's the way it works with nuget. The package download part is abstrabed away to the background.

**Comment by sylveon** [comment]
Did you install vcpkg with VS installer? That should make it work without needing to manually configure it.

![image](https://github.com/user-attachments/assets/2b75aed5-264d-4301-b350-2db558a7fcad)


FWIW NuGet for C++ doesn't restore automatically on build until you install the targets via the installer manually too

**Comment by amithegde** [comment]
yes, I have it installed and I initially follwed these instructions: https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-msbuild?pivots=shell-cmd

On the new machine, I did  not clone the vcpkg repo, I guess I will ahve to clone the repo and perform the setup on every machine. I was looking for a self contained project setup.

**Comment by sylveon** [comment]
On a brand new OS, I checked vcpkg in the VS installer (not a manual clone & integrate) and was able to build a project that uses vcpkg by just hitting F5

**Comment by amithegde** [comment]
Here is my project, please let me know how to configure it: https://github.com/amithegde/Detour

**Comment by amithegde** [comment]
ok, finally I was able to get it working. I had to remove the `vcpkg-configuration.json` and setup vcpkg as mentioend here: https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-msbuild?pivots=shell-cmd
Additionally I removed files under `vcpkg_installed` and now building on VS downloads the required files and then builds the project

**Comment by theXappy** [comment]
**tl;dr** I downloaded the masquerading NuGet package mentioned in this thread and (sort of) confirmed it was not malicious.

Regarding the aforementioned unofficial NuGet package ([here](https://www.nuget.org/packages/Detours)).
First of all, it's unlisted now (as it should be):  
![Image](https://github.com/user-attachments/assets/5a2c95da-4cff-4e91-8b3c-203a385992d5)
Here's the uploader's page: [tsungkang](https://www.nuget.org/profiles/tsungkang).  
As I've used this package before, I was afraid it contained malware. I'm leaving this comment here in case someone else mistook it for an official one and is now worried as I was.  
AFAIK no one else seem to have mention it online/acknowledge this author as malicious.

The version I had cached on disk was `Detours.4.0.1.nupkg`.
I've uploaded the nupkg to VirusTotal: https://www.virustotal.com/gui/file/178848d6d13ad1a83ced5a3aa24e73d76a5f113644e151b9c43ba132999f9118
It found nothing but I didn't trust it.

I then tried reversing it myself.  
I won't bore you all with all the details but here's the summary:
1. The `build\native\Detours.targets` file inside the nupkg [*could have*](https://github.com/augustoproiete/i-am-root-nuget-package) malicious code executable by VS on installation, but it was benign. Just copying the detours `.lib` & setting it as "additional dependency" in the project's settings.  
2. The compiled files inside are: `detours.lib` and `syelog.lib` twice, for both `x86` and `x64`.
3. Both `.lib`s have a "modified" timestamp of ` ‎7 ‎May ‎2019`. So I had to figure out the commit they used to compile them, as the repo now has newer commits. I believe this is the right commit: `d319d527a76d914610cb57f9526dc87d2703030e`
4. After compiling* the commit from (3), I used IDA to create `.idb` files for every module inside the `.lib`s (from mine + the "suspicious" version). I then loaded them into BinDiff to look for modified code.  
![Image](https://github.com/user-attachments/assets/82469bb2-3d0f-4aa5-86c6-15a70fc32f47)
As far as I can tell, then only changes were added/removed `nop` instructions (for alignment maybe? I'm not a compilers expect). For example:
![Image](https://github.com/user-attachments/assets/a83b1257-397f-47a2-9df5-700cd08b38a2)
Meaning I failed to compile them to be binarily-same, but I believe they're logically the same, making them benign.
I hope this help reassure others in my situation.**

\* I only tested the `x64` versions
** Since I'm a random guy from the internet, I'll understand if you don't blindly trust me.  
I might be working with said "tsungkang" to cover his tracks or I might just be a bad reverse engineer and missed something.  
If you installed the nupkg yourself and you're worried about infecting your PC, I suggest you either follow my steps yourself or ask an expert for help.

**Comment by sylveon** [comment]
@DrusTheAxe revisiting this: nowadays, vcpkg is preinstalled with VS and with manifest mode, you can drop a vcpkg-configuration.json inside a project to specify vcpkg dependencies to install. They will be automatically installed at build time. Perhaps this solution can work better instead of needing to vendor Detours within WASDK? I can try a PR to WASDK if you would consider it.

**Comment by sylveon** [comment]
Installed vcpkg packages are stored per-project in `<project_dir>\vcpkg_installed`, so it's similar to NuGet.

I agree the lack of a newer release is an issue, and this stems from a lack of investment in Detours from MS. I am powerless to fix that, but perhaps you can ;)
